### ISSUE TITLE
[feature] Add Support for deploying contracts with `linkReferences`

### ISSUE THREAD

**Issue by shekohex** [gakonst/ethers-rs#2296]
**Is your feature request related to a problem? Please describe.**
 
While trying to deploy a contract that contains a `linkReferences` to other libraries, the generated Rust code does not contain the bytecode for that contract; hence there was no `deploy` method. The main reason behind this is that the `bytecode` contains invalid hex; waiting for the linking to happen first, so I assume we need a preprocessing/linking support. 

**Describe the solution you'd like**
It should work as expected, the bytecode can be used as-is, later while doing the deployment, if the bytecode contains linking parts, and we have `linkReferences`, we should generate `link` method where it asks for the references first, and then we can call the `deploy` method. Or even create a `link_and_deploy` method in that case.

**Describe alternatives you've considered**

None.

**Additional context**

I would like to reference the following:
1. Original Contract code: https://github.com/webb-tools/protocol-solidity/blob/59585c2737755c80b794dc0b0555dd1dd2f8bd04/packages/contracts/contracts/hashers/PoseidonHasher.sol
2. The Generated JSON ABI: https://github.com/webb-tools/webb-rs/blob/dc7cd57afde810affd03cb57efec91085f156f43/contracts/protocol-solidity/PoseidonHasher.json
3. The Generated Rust bindgen: https://github.com/webb-tools/webb-rs/blob/dc7cd57afde810affd03cb57efec91085f156f43/src/evm/contract/protocol_solidity/poseidon_hasher.rs
4. Original Example using ethers.js from Typescript: https://github.com/webb-tools/protocol-solidity/blob/59585c2737755c80b794dc0b0555dd1dd2f8bd04/packages/anchors/src/PoseidonHasher.ts#L56-L81

**Comment by mattsse** [comment]
I guess the most straightforward approach would be to move `bytecode`

https://github.com/gakonst/ethers-rs/blob/master/ethers-solc/src/artifacts/bytecode.rs

to `core` so we can use this during abi deserialization and get linking support.

**Comment by shekohex** [comment]
Hello, is there any interest into building this? I may try to open a PR but maybe @mattsse got a better idea since looks like the `bytecode` mod uses a lot of different other modules and idk if we should also move these to the core!

**Comment by shekohex** [comment]
For now, I was able to get around this by using a hacky-way to get it to work.
TL;DR: parse the json output as `HardhatArtifact` then convert it to `ContractBytecode` later on you can do `bytecode.link_all_fully_qualified([...])` then create the `ContractFactory` yourself.

https://github.com/webb-tools/webb-rs/blob/13390b931834e6890676ee9605f064a893cb6eb5/src/evm/contract/protocol_solidity/mod.rs#L35-L86
